### PR TITLE
Make tests in `test/test_ops.py` pass for Python emulator

### PR DESCRIPTION
**Pull request by obadakhalili**
Can I lock the bounty with this PR?

Notes:
- Right now only 6 tests are failing and most of them are due to only one error. So I figured I could push a PR to lock the bounty, and in the meanwhile, fix the remaining tests.
- I left a few comments on the code which I would appreciate an answer for.

**UPDATE**: There are only 3 remaining failing tests.
**UPDATE**: All tests pass now.

### CODE REVIEW

File: tinygrad/runtime/ops_python.py [hunk 1]
```diff
@@ -107,7 +112,7 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
           else:
             ul[i] = [arg] * warp_size
         elif uop is UOps.LOOP:
-          if i not in ul:
+          if i not in ul or ul[i][0] == inp[1][0]:
             ul[i] = [0] * warp_size
```
Comment (obadakhalili) [inline tinygrad/runtime/ops_python.py:116]: This doesn't seem to be breaking any test, but shouldn't the initial value be in `inp[0][0]`?

**Comment by obadakhalili** [comment]
Regarding the `if` block in [line 120](https://github.com/tinygrad/tinygrad/pull/3384/files#diff-00bb76cdab6aa6dccea93689ae8dc2eb92343681f52231909495af017800628aR120), shouldn't this block be intended back to also check after initialization of loop state, in case the loop shouldn't be entered in the first place. Note though that this also doesn't seem to break any test. Also, implementing my suggestion would require containing the `loop_ends` in a different way.

File: tinygrad/runtime/ops_python.py [hunk 2]
```diff
@@ -82,18 +85,20 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
           loop_ends[idp[0]] = i
           i = idp[0]
           continue
-        elif uop is UOps.BARRIER:
+        # TODO: implement UOps.IF and UOps.ENDIF
+        elif uop in (UOps.BARRIER, UOps.IF, UOps.ENDIF):
```
Comment (obadakhalili) [inline tinygrad/runtime/ops_python.py:89]: I had to skip `UOps.IF` and `UOps.ENDIF` temporarily similar to `UOps.BARRIER` so I could focus on another fix, though to my surprise this didn't cause any test to fail. So I pushed this for now while I investigate it further.

File: tinygrad/runtime/ops_python.py [hunk 3]
```diff
@@ -82,18 +85,20 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
...
+
         assert dtype is not None, f"{uop} is missing a dtype"
         dl[i] = dtype
         if uop is UOps.DEFINE_GLOBAL:
           assert dtype.fmt is not None
           ul[i] = [pbufs.pop(0).cast(dtype.fmt)] * warp_size
         elif uop is UOps.DEFINE_LOCAL:
           assert dtype.fmt is not None
-          lbuf = memoryview(bytearray(arg[1]*dtype.sz))
+          lbuf = memoryview(bytearray(arg[1]*dtype.itemsize))
```
Comment (geohot) [inline tinygrad/runtime/ops_python.py:101]: This fix is definitely correct, would merge this alone right away.

**Comment by geohot** [comment]
Can you add test_ops to CI?

**Comment by obadakhalili** [comment]
@geohot I identified 3 tests that make the CI run for a long time when using the Python backend. Together with all the tests they take 1 hour to run on a personal device, and ~20 minutes without these 3 tests. I can skip them when `PYTHON=1` if you want.

**Comment by obadakhalili** [comment]
I skipped the 3 tests for now so you can inspect the rest. There are 3 remaining failing tests, which I'll start working on tomorrow morning.

**Comment by geohot** [comment]
Tests failing

**Comment by obadakhalili** [comment]
All tests pass now @geohot. And a lot of `test/test_dtype.py` tests pass too.

File: tinygrad/runtime/ops_python.py (hunk 3 above)
Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py:101]: Fixes commits are mostly independent, so I can cherry-pick this and push it as a separate PR if you want to. What about other fixes?

**Comment by geohot** [comment]
Bounty locked

File: tinygrad/runtime/ops_python.py (hunk 2 above)
Comment (geohot) [reply tinygrad/runtime/ops_python.py:89]: Yea, it's actually okay to not have IF/ENDIF, it's an optional performance improvement for GPU, not a requirement.

BARRIER isn't needed because it's always in sync in this interpreter

File: tinygrad/runtime/ops_python.py (hunk 1 above)
Comment (geohot) [reply tinygrad/runtime/ops_python.py:121]: Oh it should be, good catch.

File: tinygrad/runtime/ops_python.py [hunk 4]
```diff
@@ -137,6 +147,8 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
                 else: ret.append(_load(m, ox*4 + oy*dtp[0].shape[1]*4 + j))
               ul[i].append(ret)
           elif dtype.sz > 1:
+            if len(inp) == 4 and dtp[3].sz > 1:
```
Comment (geohot) [inline tinygrad/runtime/ops_python.py:150]: Hmm, this looks sketchy.

File: tinygrad/runtime/ops_python.py [hunk 5]
```diff
@@ -82,32 +85,39 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
           loop_ends[idp[0]] = i
           i = idp[0]
           continue
-        elif uop is UOps.BARRIER:
+        # TODO: implement UOps.IF and UOps.ENDIF
+        elif uop in (UOps.BARRIER, UOps.IF, UOps.ENDIF):
           # in the python emulator, the warp is always in sync
           i += 1
           continue
+
```
Comment (geohot) [inline tinygrad/runtime/ops_python.py:93]: Don't change whitespace if you don't have to

File: tinygrad/runtime/ops_python.py (hunk 2 above)
Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py:89]: I see. Thanks for clarifying this.

File: tinygrad/runtime/ops_python.py (hunk 1 above)
Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py:121]: - This is necessary for nested loops. Without it, you will enter the nested loop for the second time, but its state will still be pointing to `inp[1][0]`. So it is necessary to initialize the loop on first entrance, and to reinitialize on first reuse.
- Thanks, I will push a comment with the fix.

**Comment by obadakhalili** [comment]
@geohot I updated `test/test_ops.py` to skip the least amount of tests in CI that keep the runtime under 5 minutes.

**Comment by geohot** [comment]
Don't spam test_ops like that, select the tests you want from the command line

File: tinygrad/runtime/ops_python.py (hunk 4 above)
Comment (geohot) [reply tinygrad/runtime/ops_python.py:148]: Yea I don't think this is right

Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py:148]: Yes, I'm looking into it right now.

Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py]: I changed this in my second to last commit. The problem I was trying to solve is when gated `UOps.LOAD` reference values that have `.sz` > 1. I addressed that by sending the `gate`/`default` that corresponds to the current `UOps.LOAD` item.

File: test/test_ops.py [hunk 6]
```diff
@@ -1373,19 +1373,21 @@ def test_maxpool2d_bigger_stride(self):
           lambda x: torch.nn.functional.max_pool2d(x, kernel_size=(2,2), stride=stride),
           lambda x: Tensor.max_pool2d(x, kernel_size=(2,2), stride=stride))
 
-  @unittest.skipIf(Device.DEFAULT == "CUDA", "CUDA fails on this")
+  @unittest.skipIf(Device.DEFAULT in ["CUDA", "PYTHON"], "CUDA fails on this, and PYTHON is too slow")
```
Comment (geohot) [inline test/test_ops.py:1376]: No need for this, these tensor sizes are stupidly big. Want to submit another PR making them smaller? Say:

(8, 2, 17, 14) or something

Comment (obadakhalili) [reply test/test_ops.py:1376]: #3417

File: tinygrad/runtime/ops_python.py [hunk 7]
```diff
@@ -100,15 +103,20 @@ def __call__(self, *bufs, global_size:Tuple[int,int,int]=(1,1,1), local_size:Tup
...
+            ul[i] = [casted_arg] * warp_size
         elif uop is UOps.DEFINE_ACC:
           if dtype.count > 1:
             ul[i] = [[arg] * warp_size for _ in range(dtype.count)]
           else:
             ul[i] = [arg] * warp_size
         elif uop is UOps.LOOP:
-          if i not in ul:
-            ul[i] = [0] * warp_size
+          if i not in ul or ul[i][0] == inp[1][0]:
```
Comment (geohot) [inline tinygrad/runtime/ops_python.py:118]: I don't understand this or

Comment (obadakhalili) [reply tinygrad/runtime/ops_python.py:118]: There are two cases that you need to (re)initialize the loop for:
- The first is when you enter the loop for the first time. This is initialization because `ul` won't have any state about this loop.
- The second is when you enter the loop again after you have finished all its iterations. This only happens in nested loops. When the nested loop executes its last iteration, the loop will be entered again because of the next iteration of the upper loop, which will be problematic because the state of the previous iteration of the nested loop is still saved in `ul` and needs to be reinitialized to `inp[0][0]`. But now while explaining it, I realize that the best solution is to clear the state of the loop after the last iteration. I will push a commit with that.

**Comment by geohot** [comment]
PayPal or USDC for your $100? E-mail george@tinygrad.org with email/address to claim.

**Comment by Max-We** [comment]
Are you sure that this is fixed? Running `PYTHON=1 python3 test/test_ops.py -f` on my machine / Google Colab throws and error. Of course, without the `PYTHON=1` flag it works, but that's not the point, no?

**Comment by iglesias** [comment]
"$100 bounties are a few line change and could be 10 minutes if you see it" 

hmm, I think someone is not being precisely candid,
